dask/dataframe/optimize.py

```diff
@@ -84,49 +92,49 @@ def optimize_dataframe_getitem(dsk, keys):
                 # ... where this value is __getitem__...
                 return dsk
 
-            if any(layers[k].name == x[0] for x in keys if isinstance(x, tuple)):
-                # ... but bail on the optimization if the dataframe_blockwise layer is in
-                # the requested keys, because we cannot change the name anymore.
-                # These keys are structured like [('getitem-<token>', 0), ...]
-                # so we check for the first item of the tuple.
-                # See https://github.com/dask/dask/issues/5893
-                return dsk
+            if block.indices[1][1] is not None:
```

Comment:
It wasn't immediately clear to me that this condition is what's needed for selecting out column projection layers. I wonder if there's something more direct we can check for here

Reply:
I agree that it would be best to have a more-intuitive check here... Still thinking about this.

Reply:
I guess this is more motivation for a formal getitem Layer...  What we are looking for here is a `Blockwise` layer, based on `operator.getitem`, with the getitem-key (`block.indices[1][0]`) being a `str` or `list`.  We do want the getitem key to be a literal (so `block.indices[1][0]` should be `None`), but the only time it will not be a literal is if the key is another layer/collection key.  Therefore, we can either check that the getitem key is not a layer key, or we can check that the indice is None.

Reply:
Thanks for the updates @rjzamora. I agree we might want to add some additional structure long-term to make this code easier to reason about. Though for now, the current set  of changes look good and fix https://github.com/dask/dask/issues/7692